stl/inc/future

```diff
@@ -285,18 +298,36 @@ public:
         return _Result;
     }
 
+    template <typename _Ty2>
+    void _Emplace_result_2(_Ty2&& _Val, true_type) {
+        _Result = _STD forward<_Ty2>(_Val);
+    }
+
+    template <typename _Ty2>
+    void _Emplace_result_2(_Ty2&& _Val, false_type) {
+        ::new (static_cast<void*>(_STD addressof(_Result._HeldValue))) _Ty(_STD forward<_Ty2>(_Val));
+        _Has_stored_result = true;
+    }
+
+    template <typename _Ty2>
+    void _Emplace_result(_Ty2&& _Val) {
+        // TRANSITION: _Associated_state ctor will move or copy assign if _Result was default constructed
+        // rather than move or copy construct
+        _Emplace_result_2(_STD forward<_Ty2>(_Val), is_default_constructible<_Ty>{});
+    }

```

Comment:
I think MSVC STL tends to reduce tag dispatch by `if constexpr` (a C++17 feature usable as an extension in C++14 mode).
```suggestion
    template <typename _Ty2>
    void _Emplace_result(_Ty2&& _Val) {
        // TRANSITION: _Associated_state ctor will move or copy assign if _Result was default constructed
        // rather than move or copy construct
        if constexpr (is_default_constructible_v<_Ty>) {
            _Result = _STD forward<_Ty2>(_Val);
        } else {
            ::new (static_cast<void*>(_STD addressof(_Result._HeldValue))) _Ty(_STD forward<_Ty2>(_Val));
            _Has_stored_result = true;
        }
    }
```

Reply:
Yes, we strongly prefer `if constexpr` and use it in all new code. Its advantages are numerous (shorter and more readable source code, faster for the compiler front-end to instantiate, faster for the compiler back-end to deal with as it emits far less codegen). There's only one situation where `if constexpr` doesn't supersede tag dispatch and it's extremely rare (as `if constexpr` happens in a function body, it can't select between different member initializer lists in a constructor).

We're gradually cleaning up our codebase to convert old uses of tag dispatch to `if constexpr`, so if you see tag dispatch in existing code, that's why.

We have globally suppressed (within STL headers) the compiler warning that using `if constexpr` in C++14 is using Future Technology, so you can use it with no further issues.